File: 13.罗马数字转整数.py
#
# @lc app=leetcode.cn id=13 lang=python3
#
# [13] 罗马数字转整数
#

# @lc code=start


class Solution:
    def romanToInt(self, s: str) -> int:
        nums = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
        romans = ["M", "CM", "D", "CD", "C", "XC",
                  "L", "XL", "X", "IX", "V", "IV", "I"]
        res = 0
        index = 0
        num_index = 0
        while index < len(s):
            while num_index < len(nums):
                while s[index:].startswith(romans[num_index]):
                    index += len(romans[num_index])
                    res += nums[num_index]
                num_index += 1
        return res
# @lc code=end

File: 13.罗马数字转整数.java
/*
 * @lc app=leetcode.cn id=13 lang=java
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
class Solution {
    public int romanToInt(String s) {
        int sum = 0;
        int cur = wordToInt(s.charAt(0));
        int next = 0;
        for (int i = 0; i < s.length() - 1; i++) {
            next = wordToInt(s.charAt(i+1));
            if (next > cur)
                sum -= cur;
            else
                sum += cur;
            cur = next;
        }
        return sum + wordToInt(s.charAt(s.length() - 1));

    }
    public int wordToInt(char ch) {
        switch(ch) {
            case 'I': return 1;
            case 'V': return 5;
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
        }
        return 0;
    }
}
// @lc code=end

